### ISSUE TITLE
Slow query - get entity by classification in JanusGraph

### ISSUE THREAD

**Issue by Mirovan** [odpi/egeria#4479]
Hi!

I want get some entities by classification.

When I call API open-metadata/repository-services/users/<user>/entities/by-classification/<slassification-name> - it is work, but very slow - 10-30 seconds.

I am try make test which get entities data without API:
```
            GraphTraversal<Vertex, Vertex> gt = g.V().hasLabel("Classification");
            g.V().has("vcclassificationName", "ParentGroupClassification");

            end = System.currentTimeMillis();

            System.out.printf("getGremlinTime: %d ms\n", (end - start));

            start = end;
            List<Vertex> res2 = gt.toList();

            end = System.currentTimeMillis();
            System.out.printf("toList: %d ms\n gremlin count: %d\n", (end - start), res2.size());
```

But I have slow result:
```
getGremlinTime: 17 ms

2021-01-11 12:55:19.834  WARN 24780 --- [           main] o.j.g.transaction.StandardJanusGraphTx   : Query requires iterating over all vertices [(~label = Classification)]. For better performance, use indexes

toList: 10451 ms
 gremlin count: 7
```

I think that this problem in - g.V().hasLabel("Classification"), and we need new index in janusgraph.
There are same error and issue of Janusgraph:
https://github.com/JanusGraph/janusgraph/issues/2324
https://github.com/JanusGraph/janusgraph/issues/283

How I can improve Egeria queries to Janusgraph ?

@grahamwallis @mandy-chessell @davidradl

**Comment by grahamwallis** [comment]
Hi @Mirovan - There is a mixed index on classification name, so if you issue a traversal that includes a classification name the index should be used. 

It looks from your example above as if you might have issued a traversal that only asked for the label to be Classification. That probably would be a slow traversal as the label is not really useful by itself; it would really just be saying "find all the classification vertices". If you use the MetadataCollection API to find entities by classification it will perform a more specific traversal.

Hope that helps.

**Comment by Mirovan** [comment]
Hi @grahamwallis 

In class GraphOMRSMetadataStore, method findEntitiesByClassification we have some code:
```
        GraphTraversal<Vertex, Vertex> gt = g.V().hasLabel("Classification");
```
I think that this line - too slow, if delete this line we have same behavior, but it will be more fast.

I create test:
```
//fast:
GraphTraversal<Vertex, Vertex> gt = g.V();
List<Vertex> res2 = gt.toList();
//330 ms

//this is slow:
GraphTraversal<Vertex, Vertex> gt = g.V().hasLabel("Classification");
List<Vertex> res2 = gt.toList();
//30 sec
```
Can we delete line in source **GraphTraversal<Vertex, Vertex> gt = g.V().hasLabel("Classification");** and replace with **GraphTraversal<Vertex, Vertex> gt = g.V();**

Is I right ?

**Comment by Mirovan** [comment]
As I know index by label - not possible in JanusGraph - https://github.com/JanusGraph/janusgraph/issues/283

**Comment by grahamwallis** [comment]
Hi @mirovan - It is worth trying :-) 

The traversal gets built up and is only executed when all the necessary steps have bee added. The later steps in the traversal are the ones where we want to hit the indexes. There is a possibility that having the label step is actually causing us to enumerate all the classification vertices, whereas cutting straight to the chase would allow us to hit an index sooner - which would be a good thing. I have not tried to do any traversal optimisation in the graph repo, so I am sure there are opportunities like this.

**Comment by grahamwallis** [comment]
Hi @Mirovan 

I did some experiments and got the following results:

The experiments used a gremlin console and a graph with a Berkeley database.

Categories:
There are two categories of vertex - person and address.

Labels:
Each vertex will have label set to ‘person’ or ‘address’.

Properties:
The person vertices have a property ‘name’. 
Every person vertex will have ‘name’ set to a unique value e.g. ‘name-1’, name-2’.
The address vertices have a property ‘postcode’. 
Every address vertex will have ‘postcode’ set to a unique value e.g. ‘postcode-1’, postcode-2’.
In both experiments we will use a composite index on ‘name’ but no index on ‘postcode’ - so we can compare indexed and non-indexed property search.

Additional ‘labelling’ property:
Every vertex also has a property called ‘ecat’, which will have the same value as the vertex label (‘person’ or ‘address’).
Using an index on ecat may not be productive because it will be a ‘hot’ index - there will be many vertices with the same 
value for ecat, so the index will have low selectivity.

Indexes:
In experiment 1, we will use a composite index on ‘ecat’ to see if it speeds up label-based traversals (using the ‘ecat’ property). A composite index is managed internally to JanusGraph.
In experiment 2, we will use a mixed index on ‘ecat’ for comparison, because the JanusGraph index documentation suggests that a mixed index is better for data with low selectivity (we only have two values for ‘ecat’). A mixed index uses an indexing backend - in this experiment it’s Lucene.

Populations:
In both experiments:
There will always be 1K person vertices.
We will increase the number of address vertices from 1K to 10K to 100K.

Experiment source code:

```
// Perform a traversal that starts with hasLabel("person") and measure
// Perform a traversal that omits the hasLabel("person") step and measure
// Perform a traversal that includes the hasLabel("person") step at the end and measure
// Compare results
//
//
// You can run this query by going into bin/gremlin.sh and typing:
//
// :load ../../gremlin/label-performance/gremlin
//

graph =  JanusGraphFactory.open("conf/janusgraph-berkeleyje-lucene.properties")


mgmt = graph.openManagement()

// To use a composite index for 'ecat' uncomment the following lines and comment the ones for mixed index below:
// Create a propertyKey and composite index for egeria category
// ecat = mgmt.makePropertyKey('ecat').dataType(String.class).make()
// mgmt.buildIndex('byEcatComposite', Vertex.class).addKey(ecat).buildCompositeIndex()

// To use a mixed index for 'ecat' uncomment the following lines and comment the ones for composite index above:
// Create a propertyKey and mixed index for egeria category
ecat = mgmt.makePropertyKey('ecat').dataType(String.class).make()
mgmt.buildIndex('byEcatMixed', Vertex.class).addKey(ecat).buildMixedIndex("search")

// Create a propertyKey and composite index for the only property in this test case
name = mgmt.makePropertyKey('name').dataType(String.class).make()
mgmt.buildIndex('byNameComposite', Vertex.class).addKey(name).buildCompositeIndex()

// You will need to commit the mgmt transaction before waiting for the index status
mgmt.commit()

// Wait for the indexes to become available
ManagementSystem.awaitGraphIndexStatus(graph, 'byEcatMixed').call()
ManagementSystem.awaitGraphIndexStatus(graph, 'byNameComposite').call()

g = graph.traversal()

// Graph should be empty...
g.V().count()
g.E().count()

// Insert vertices - each does have a label but we also set the (indexed) ecat category property
// This risks creating a hot index (every entity will have the same valie, every edge, etc)
// but as an experiment it is worth setting this and testing the query performance wit it.

for (i=0; i<1000; i++) {
  g.addV('person').property('ecat','person').property('name','number-'+i).next()
}

for (i=0; i<100000; i++) {
  g.addV('address').property('ecat','address').property('postcode','postcode-'+i).next()
}

// Commit the vertex creations - otherwise they will not show up in the mixed index backend
g.tx().commit()

//
// Counts should be 2000 and 0
g.V().count()
g.E().count()


// Perform a trivial query 
g.V().has('name','number-6')

// Now enter manual traversals.....
```

On a clean system (no persistent store or indexes), I start a gremlin console and load the above source file.

I then run the following traversals capturing the profile of each - to check it is using the index we expect and to measure 
how long it takes to run. Each traversal is run multiple times (e.g. x10) and record the median time.
```
g.V()                                                 
g.V().hasLabel('person')                 
g.V().has(‘ecat’,’person’)                 
g.V().has(‘name’,’number-563’)  
g.V().hasLabel(‘address’)                 
g.V().has(‘ecat’,’address’)                 
g.V().has(‘postcode’,’postcode-211’)  
```
When the mixed index is used I have to modify the ‘ecat’ traversals slightly to ensure the query uses the mixed index. 
So they become:
```
g.V().has(‘ecat’,textContains(’person’))
g.V().has(‘ecat’,textContains(’address’))
```

The results look like this:

Composite Index

![image](https://user-images.githubusercontent.com/17125304/105400326-af9be000-5c1c-11eb-9b74-a0dc36e9af27.png)

Mixed Index

![image](https://user-images.githubusercontent.com/17125304/105400356-b88cb180-5c1c-11eb-8fab-ed7c9f1ca286.png)


My observations are:

1. The traversals that perform full scans get linearly slower the more vertices they retrieve. This is obvious really,
but it’s worth reflecting on which traversals these are: g.V() is an obvious one, but also the hasLabel() step (alone)
does not use an index, nor do the traversals retrieving based on a non-indexed property (‘postcode’). All of these 
traversals show heavy linear growth in execution time on the charts. 

2. A traversal that retrieves based on an indexed property (‘name’) shows constant performance despite the growing 
population of ‘other’ vertices. This is true with both the composite and mixed indexes.

3. The use of an indexed property instead of the label - i.e using has(‘ecat’,’person’) vs hasLabel(‘person’) shows a definite benefit for vertices that have relatively small populations. In the charts, notice that person vertices retrieved using has(‘ecat’,’person’) showed almost constant performance, with either type of index. There is a noticeable benefit of the mixed index in teh test where there were 1K person vertices and 1K address vertices. Initially I thought this was an anomaly, but the same behavior is seen for has(‘ecat’,’person’) and has(‘ecat’,’address’) across both index runs. So it appears to be a consistent and repeatable artefact. This suggests that for the more balanced population (1K/1K) the mixed index offers an advantage over the composite index. For the more biased populations the larger (address) population shows a marked degradation as more vertices are retrieved, but that is probably to be expected.

Applying this back to the Egeria graph performance discussion:

There is obvious benefit in using a property as a substitute for a label for small populations of vertices. There is far less benefit for larger populations. If we were to query Egeria to retrieve 1000 classifications from a repository that also contains 100000 entities, then we should see the performance benefit exhibited by the has(‘ecat’,’person’) results
in the experiments. 

However, the above would only provide benefit where the population is relatively small (due to the selectivity of the index) and where no other index lookups are being performed. i.e. a traversal such as g.V().hasLabel(‘Classification’) would be better run as g.V().has(‘ecat’,’Classification’). However, it’s not clear (to me, yet) that such a traversal is actually useful. It’s not a traversal that the metadata collection interface really encourages. If a user wishes to search for entities by classification (using the findEntitiesByClassification method) they must provide the classificationName - and that resultsin an additional indexed property being included in the traversal. In JanusGraph the traversal:

g.V().hasLabel(‘Classification’).has(‘name’,’name-66’) is parsed into a single step and executed as has(label,key,value) and uses the index, as follows:
```
gremlin> g.V().hasLabel('person').has('name', 'number-33').profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[~label.eq(person), name.eq(n...                     1           1           0.958   100.00
    \_condition=(~label = person AND name = number-33)
    \_orders=[]
    \_isFitted=false
    \_isOrdered=true
    \_query=multiKSQ[1]@2000
    \_index=byNameComposite
  optimization                                                                                 0.013
  optimization                                                                                 0.272
  backend-query                                                        1                       0.539
    \_query=byNameComposite:multiKSQ[1]@2000
    \_limit=2000
                                            >TOTAL                     -           -           0.958        -
```

I am not sure therefore that in an Egeria traversal, there would be any performance boost from the introduction of an additional property to enable index lookup of the label value. I am not aware of any traversals in the graph repository that use the hasLabel() step and do not have additional steps that refer to indexed properties, such as GUID, typeDefName, classificationName, etc.

If we dropped the hasLabel() step, or an equivalent property based step from the traversals then as far as I am aware the 
traversals would all still deliver correct results, due to the presence of those additional properties and the design of the graph schema in which each category of vertex or edge uses a distinct property name prefix. It would introduce a risk that if, in future, a less-specific traversal were introduced, it may capture incorrect results.

**Comment by Mirovan** [comment]
Hi @grahamwallis !

I am agree with you. In gremlin console I have same result. But when I create Java-code I have very slow select result.
Now I find solve for this problem. I rewrite some method in my application. Insert query - I use Egeria API, select query - I use my custom code, like this:
```
    public List<EntityDetail> findEntitiesByKeyValue(String key, String value) {
        List<EntityDetail> entities = null;

        try (GraphTraversalSource g = instanceGraph.traversal()) {
            //поиск всех смежных входящих вершин по ключу и значению - label, key, value
            GraphTraversal<Vertex, Vertex> gt = g.V().has(key, value).in();

            entities = new ArrayList<>();

            while (gt.hasNext()) {
                Vertex entityVertex = gt.next();

                EntityDetail entityDetail = vertexToEntity(entityVertex);
                try {
                    entities.add(entityDetail);
                } catch (Exception e) {
                    continue; // process the next vertex
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        return Optional
                .ofNullable(entities)
                .orElseThrow(()-> new RuntimeException("Ошибка получения терминов-родителей 0-го уровня с классификатором 'ParentGroupClassification'."));
    }


    /**
     * Поиск смежных вершин по связи Relationship
     * @param guid - ID вершины
     * */
    public List<EntityDetail> findAdjacentEntities(String guid) {
        List<EntityDetail> entities = null;

        try (GraphTraversalSource g = instanceGraph.traversal()) {
            //поиск всех смежных входящих вершин по ключу и значению - label, key, value
            GraphTraversal<Vertex, Vertex> gt =
                    g.V()
                            .has("veguid", guid)
                            .outE()
                            .has(T.label, "Relationship")
                            .inV();

            entities = new ArrayList<>();

            while (gt.hasNext()) {
                Vertex entityVertex = gt.next();

                EntityDetail entityDetail = vertexToEntity(entityVertex);
                try {
                    entities.add(entityDetail);
                } catch (Exception e) {
                    continue; // process the next vertex
                }
            }

        } catch (Exception e) {
            e.printStackTrace();
        }

        return Optional
                .ofNullable(entities)
                .orElseThrow(()-> new RuntimeException("Ошибка получения терминов-родителей 0-го уровня с классификатором 'ParentGroupClassification'."));
    }
```

For my problem - I want get all Entities with zero-level. I create new Classification and create relationship from zero-level Entity to Classification:
```
List<EntityDetail> entities = findEntitiesByKeyValue("vcclassificationName", "ParentGroupClassification");
```

**Comment by grahamwallis** [comment]
Hi @Mirovan 

I did a little more investigation into this on Friday. Based on your comment that you thought hasLabel() was slowing down the query, and that you had experienced differences between gremlin console and Java, I thought I should perform an experiment using the real Egeria graph repository. So I added timing debug statements and experimented with removing the hasLabel('Classification') step from the implementation of findEntitiesByClassification in the metadata store class.

I then ran two sets of test - with x 10 repeats in each test. In one test, I included the hasLabel() step, and in the other I had removed it. Both sets of tests were run from a clean start - i.e. restart of the platform and start of the server to pick up an existing graph repository that contains a number of classified entities.

The results were quite 'noisy' - but I didn't see any consistent or marked advantage whether the label step is included or not - which I think is consistent with the results of my earlier gremlin experiments and reflect the use of the property index for the classification name property.

<img width="361" alt="image" src="https://user-images.githubusercontent.com/17125304/105691700-7e116600-5ef5-11eb-8ac0-38be726d6831.png">

Unfortunately I can't strip the graph repository implementation down to the same extent that you have (above in your findEntitiesByKeyValue method) because it needs to all the options in the metadata collection API - including support type filtering and property filtering, but I think I can reorganise it to be more efficient. I'm going to look at restructuring the delegation of the entity type filtering from the metadata collection class into the metadata store class, so I can perform the filtering within the traversal. The current delegation was based on the implementation of the other find methods and is not efficient for classification based searches.

By the way - one thing I should mention is that whilst it is possible to write methods (as you have above) based on the current graph schema. be aware that the schema could change - whereas the metadata collection interface should retain backward compatibility. If there are additional functions that would be useful we should look at adding them to the interface.

I will let you know how today's restructuring goes.

Best regards, Graham

**Comment by Mirovan** [comment]
Thank you Graham.
Now we are using Egeria 2.5. We try use your investigation later )
It good answer for me!
Thank you!

**Comment by grahamwallis** [comment]
Hopefully this issue is addressed by PR #4523.

**Comment by grahamwallis** [comment]
Further performance updates merged in #4684
